client/emv/emvcore.c

```diff
@@ -23,11 +94,370 @@ void TLVPrintFromBuffer(uint8_t *data, int datalen) {
+}
+
+void TLVPrintAIDlistFromSelectTLV(struct tlvdb *tlv) {
+	PrintAndLog("|------------------|--------|-------------------------|");
+	PrintAndLog("|    AID           |Priority| Name                    |");
+	PrintAndLog("|------------------|--------|-------------------------|");
+
+	struct tlvdb *ttmp = tlvdb_find(tlv, 0x6f);
+	if (!ttmp)
+		PrintAndLog("|                         none                        |");
+		
+	while (ttmp) {
+		const struct tlv *tgAID = tlvdb_get_inchild(ttmp, 0x84, NULL);
+		const struct tlv *tgName = tlvdb_get_inchild(ttmp, 0x50, NULL);
+		const struct tlv *tgPrio = tlvdb_get_inchild(ttmp, 0x87, NULL);
+		if (!tgAID)
+			break;
+		PrintAndLog("|%s|   %s  |%s|", 
+			sprint_hex_inrow_ex(tgAID->value, tgAID->len, 18), 
+			(tgPrio) ? sprint_hex(tgPrio->value, 1) : "   ", 
+			(tgName) ? sprint_ascii_ex(tgName->value, tgName->len, 25) : "                         ");
+		
+		ttmp = tlvdb_find_next(ttmp, 0x6f);
+	}
+
+	PrintAndLog("|------------------|--------|-------------------------|");
+}
+
+
+int EMVSelect(bool ActivateField, bool LeaveFieldON, uint8_t *AID, size_t AIDLen, uint8_t *Result, size_t MaxResultLen, size_t *ResultLen, uint16_t *sw, struct tlvdb *tlv) {
+	uint8_t data[260] = {0};
+	*ResultLen = 0;
+	if (sw)	*sw = 0;
+	uint16_t isw = 0;
+	
+	// select APDU
+	data[0] = 0x00;
+	data[1] = 0xA4;
+	data[2] = 0x04;
+	data[3] = 0x00;
+	data[4] = AIDLen;
+	memcpy(&data[5], AID, AIDLen);
+	
+	if (ActivateField)
+		DropField();
+	
+	if (APDULogging)
+		PrintAndLog(">>>> %s", sprint_hex(data, AIDLen + 5));
+
+	int res = ExchangeAPDU14a(data, AIDLen + 5, ActivateField, LeaveFieldON, Result, (int *)ResultLen);
+	
+	if (APDULogging)
+		PrintAndLog("<<<< %s", sprint_hex(Result, *ResultLen));
+	
+	if (res) {
+		return res;
+	}
+	
+	if (*ResultLen < 2) {
+		PrintAndLog("SELECT ERROR: returned %d bytes", *ResultLen);
+		return 5;
+	}
+	
+	*ResultLen -= 2;
+	isw = Result[*ResultLen] * 0x0100 + Result[*ResultLen + 1];
+	if (sw)
+		*sw = isw;
+
+	if (isw != 0x9000) {
+		if (APDULogging)
+			PrintAndLog("SELECT ERROR: [%4X] %s", isw, GetAPDUCodeDescription(*sw >> 8, *sw & 0xff));
+		return 5;
+	}
+
+	// add to tlv tree
+	if (tlv) {
+		struct tlvdb *t = tlvdb_parse_multi(Result, *ResultLen);
+		tlvdb_add(tlv, t);
+	}
+	
+	return 0;
+}
+
+int EMVSelectPSE(bool ActivateField, bool LeaveFieldON, uint8_t PSENum, uint8_t *Result, size_t MaxResultLen, size_t *ResultLen, uint16_t *sw) {
+	uint8_t buf[50] = {0};
+	*ResultLen = 0;
+	int len = 0;
+	int res = 0;
+	switch (PSENum) {
+		case 1:
+			param_gethex_to_eol(PSElist[1], 0, buf, sizeof(buf), &len);
+			break;
+		case 2:
+			param_gethex_to_eol(PSElist[0], 0, buf, sizeof(buf), &len);
+			break;
+		default:
+			return -1;
+	}
+	
+	// select
+	res = EMVSelect(ActivateField, LeaveFieldON, buf, len, Result, MaxResultLen, ResultLen, sw, NULL);
+
+	return res;
+}
+
+int EMVSearchPSE(bool ActivateField, bool LeaveFieldON, bool decodeTLV, struct tlvdb *tlv) {
+	uint8_t data[260] = {0};
```

Comment:
Why 260?

Reply:
fixed. 50 bytes = 100hex symbols. it enough for AID (ususally they have only 7-10 bytes)

Reply:
fixed

Reply:
No?